massa-consensus-worker/src/state/blocks_state.rs

```diff
@@ -0,0 +1,289 @@
+use std::collections::hash_map::Entry;
+
+use massa_consensus_exports::{block_status::BlockStatus, error::ConsensusError};
+use massa_models::{
+    block_id::BlockId,
+    prehash::{PreHashMap, PreHashSet},
+    slot::Slot,
+};
+
+#[derive(Debug, Clone)]
+pub struct BlocksState {
+    /// Every block we know about
+    block_statuses: PreHashMap<BlockId, BlockStatus>,
+    /// Ids of incoming blocks/headers
+    incoming_index: PreHashSet<BlockId>,
+    /// Used to limit the number of waiting and discarded blocks
+    sequence_counter: u64,
+    /// ids of waiting for slot blocks/headers
+    waiting_for_slot_index: PreHashSet<BlockId>,
+    /// ids of waiting for dependencies blocks/headers
+    waiting_for_dependencies_index: PreHashSet<BlockId>,
+    /// ids of discarded blocks
+    discarded_index: PreHashSet<BlockId>,
+    /// ids of active blocks
+    active_index: PreHashSet<BlockId>,
+}
+
+impl BlocksState {
+    pub fn new() -> Self {
+        Self {
+            block_statuses: PreHashMap::default(),
+            incoming_index: PreHashSet::default(),
+            sequence_counter: 0,
+            waiting_for_slot_index: PreHashSet::default(),
+            waiting_for_dependencies_index: PreHashSet::default(),
+            discarded_index: PreHashSet::default(),
+            active_index: PreHashSet::default(),
+        }
+    }
+
+    pub fn get(&self, block_id: &BlockId) -> Option<&BlockStatus> {
+        self.block_statuses.get(block_id)
+    }
+
+    pub fn get_mut(&mut self, block_id: &BlockId) -> Option<&mut BlockStatus> {
+        self.block_statuses.get_mut(block_id)
+    }
+
+    pub fn sequence_counter(&self) -> u64 {
+        self.sequence_counter
+    }
+
+    pub fn inc_sequence_counter(&mut self) {
+        self.sequence_counter += 1
+    }
+
+    pub fn incoming_blocks(&self) -> &PreHashSet<BlockId> {
+        &self.incoming_index
+    }
+
+    pub fn waiting_for_slot_blocks(&self) -> &PreHashSet<BlockId> {
+        &self.waiting_for_slot_index
+    }
+
+    pub fn waiting_for_dependencies_blocks(&self) -> &PreHashSet<BlockId> {
+        &self.waiting_for_dependencies_index
+    }
+
+    pub fn discarded_blocks(&self) -> &PreHashSet<BlockId> {
+        &self.discarded_index
+    }
+
+    pub fn active_blocks(&self) -> &PreHashSet<BlockId> {
+        &self.active_index
+    }
+
+    // Internal function to update the indexes
+    fn update_indexes(
```

Comment:
maybe this could be changed into: 
```
transition_block_state(&mut self, block_id: &BlockId, new_status: Option<&BlockStatus>) {
   // here we automatically go check the old state (if present), verify that the transition is allowed, and perform the transition, together with updating the indexes 
}


```

Reply:
Yes it seems the same as directly change the index in `update_block_state` I will try

Reply:
I'm not a big fan of the cloning of Active blocks because it clones the Storage of the block, taking locks on block storage, endorsement storage and operation storage, and taking 5000 references to operations. Then it happens again when the cloned object is dropped

Reply:
So if we imagine coming back to the "removal" pattern, we can do:

```rust
pub fn transition_map(&mut self, block_id: &BlockId, Fn map_fn(Option<BlockState>)->Option<BlockState>) {
   // removes the block from block_statuses, getting its current state in the process and updating indexes
   // calls map_fn that consomes the old state, and returns the new state
   // add the new state (if any) to block_statuses and updating indexes
 }
```

Reply:
I have updated the only clone that is left is : https://github.com/massalabs/massa/blob/c6949ca9f1a045a1428424eedb5c7379f534f465/massa-consensus-worker/src/state/process.rs#L158 and it's a pain to remove

Reply:
Removed. I should be clean now. Labnet test running.